### PR TITLE
Remove #[doc(hidden)] from stable crates

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
This PR removes `#[doc(hidden)]` from types/functions in stable crates (as defined in [CrateSet.kt](https://github.com/smithy-lang/smithy-rs/blob/ad520b080aa0a3e8239ab2857dafb519c3458088/buildSrc/src/main/kotlin/CrateSet.kt#L19-L33)). They are now `pub`, however, for those that are not intended to be used directly, we preserve the original docs to be explicit about it.

## Testing
After this PR, stable crates contain neither
<details>
<summary>#[doc(hidden)]</summary>

```
➜  smithy-rs git:(ysaito/remove-doc-hidden) rg -l '#\[doc\(hidden\)\]'
aws/rust-runtime/aws-http/src/user_agent.rs
CHANGELOG.next.toml
CHANGELOG.md
rust-runtime/aws-smithy-http/src/lib.rs
rust-runtime/aws-smithy-http/src/event_stream/sender.rs
rust-runtime/aws-smithy-http/src/event_stream/receiver.rs
codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/error/ErrorImplGenerator.kt (this is only for CodegenTarget.SERVER)
codegen-core/src/test/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/StructureGeneratorTest.kt
design/src/server/instrumentation.md
design/src/rfcs/rfc0039_forward_compatible_errors.md
design/src/rfcs/rfc0032_better_constraint_violations.md
design/src/rfcs/rfc0026_client_crate_organization.md
rust-runtime/aws-smithy-http-server/src/macros.rs
rust-runtime/aws-smithy-http-server/src/routing/mod.rs
rust-runtime/aws-smithy-http-server/src/shape_id.rs
rust-runtime/aws-smithy-http-server/src/body.rs
rust-runtime/aws-smithy-http-server/src/lib.rs
rust-runtime/aws-smithy-http-server/src/plugin/mod.rs
codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/MapConstraintViolationGenerator.kt
codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/CollectionConstraintViolationGenerator.kt
```
</details>

nor 

<details>
<summary>DocHidden</summary>

```
➜  smithy-rs git:(ysaito/remove-doc-hidden) rg -l 'DocHidden'         
codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustType.kt
codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderConstraintViolations.kt
```
</details>

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/rust-runtime/aws-inlineable/src/s3_request_id.rs [hunk 1]
```diff
@@ -70,7 +70,6 @@ where
 }
 
 /// Applies the extended request ID to a generic error builder
-#[doc(hidden)]
 pub fn apply_extended_request_id(
```
Comment (jdisanti) [inline aws/rust-runtime/aws-inlineable/src/s3_request_id.rs:73]: I think this one can be `pub(crate)`

File: rust-runtime/aws-smithy-runtime-api/src/client/interceptors/context.rs [hunk 2]
```diff
@@ -222,7 +222,6 @@ impl<I, O, E> InterceptorContext<I, O, E> {
     }
 
     /// Advance to the Serialization phase.
-    #[doc(hidden)]
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime-api/src/client/interceptors/context.rs:225]: All these interceptor context ones should say they're not intended to be used

File: rust-runtime/aws-smithy-runtime-api/src/lib.rs [hunk 3]
```diff
@@ -33,8 +33,7 @@ pub mod client;
 
 pub mod http;
 
-/// Internal builder macros. Not intended to be used outside of the aws-smithy-runtime crates.
-#[doc(hidden)]
+/// Builder macros. Not intended to be used outside of the aws-smithy-runtime crates.
 pub mod macros;
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime-api/src/lib.rs:37]: I think we should expand these macros where they're used and delete them. They're not fully baked enough to be stable.

**Review by jdisanti — COMMENTED** [review]
Great work on this! Just a few things.

File: aws/rust-runtime/aws-inlineable/src/s3_request_id.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply aws/rust-runtime/aws-inlineable/src/s3_request_id.rs]: Addressed in [6433ca](https://github.com/smithy-lang/smithy-rs/pull/3226/commits/c6433ca7f2c351cf27a5c2570a32db0538e363e8).

Comment (ysaito1001) [reply aws/rust-runtime/aws-inlineable/src/s3_request_id.rs]: Addressed in [6433ca](https://github.com/smithy-lang/smithy-rs/pull/3226/commits/c6433ca7f2c351cf27a5c2570a32db0538e363e8).

File: rust-runtime/aws-smithy-runtime-api/src/lib.rs (hunk 3 above)
Comment (ysaito1001) [reply rust-runtime/aws-smithy-runtime-api/src/lib.rs]: Great call out. Updated in [55f1c75](https://github.com/smithy-lang/smithy-rs/pull/3226/commits/55f1c75559c80281eef9cac17db3f54f253d7cc7).

File: rust-runtime/aws-smithy-runtime-api/src/client/interceptors/context.rs (hunk 2 above)
Comment (ysaito1001) [reply rust-runtime/aws-smithy-runtime-api/src/client/interceptors/context.rs:225]: Good call out. Addressed in [55f1c75](https://github.com/smithy-lang/smithy-rs/pull/3226/commits/55f1c75559c80281eef9cac17db3f54f253d7cc7).

**Review by jdisanti — APPROVED** [review]
🚀

File: rust-runtime/aws-smithy-runtime-api/src/lib.rs (hunk 3 above)
Comment (ysaito1001) [reply rust-runtime/aws-smithy-runtime-api/src/lib.rs]: Good catch. Currently this module is used within a single module in `aws-smithy-rust-runtime`. Makes sense to inline the expanded code in that module and delete `macros` at least for now. Done in [3e9bcc6](https://github.com/smithy-lang/smithy-rs/pull/3226/commits/3e9bcc6313907fc8100fcddbb8436f2e901d85f4).

**Pull request merged**